services/streammanager/googlesheets/googlesheetsmanager.go

```diff
@@ -59,67 +52,94 @@ func init() {
 	}
 
 	var opts []option.ClientOption
 	if config.TestConfig.Endpoint != "" { // test configuration
 		opts = testClientOptions(&config)
-	} else { // normal configuration
-		if opts, err = clientOptions(&config); err != nil {
+	} else {
+		jwtConfig, err := google.JWTConfigFromJSON([]byte(config.Credentials), sheets.SpreadsheetsScope)
+		if err != nil {
 			return nil, fmt.Errorf("[GoogleSheets] error :: %w", err)
 		}
+
+		oauth2Client := oauth2.NewClient(ctx, jwtConfig.TokenSource(ctx))
+		opts = append(opts, option.WithHTTPClient(oauth2Client))
 	}
 
-	service, err := generateService(opts...)
+	service, err := sheets.NewService(ctx, opts...)
 	// If err is not nil then retrun
 	if err != nil {
 		pkgLogger.Errorn("[Googlesheets] error", obskit.Error(err))
 		return nil, err
 	}
 
+	return &GoogleSheetsProducer{
+		config:          config,
+		client:          &Client{service, o},
+		lock:            sync.Mutex{},
+		isHeaderUpdated: false,
+	}, err
+}
+
+func (p *GoogleSheetsProducer) updateHeader() error {
+	if p.isHeaderUpdated {
+		return nil
+	}
+	p.lock.Lock()
+	defer p.lock.Unlock()
+	if p.isHeaderUpdated {
+		return nil
+	}
+
+	var headerRowStr []string
 	// ** Preparing the Header Data **
 	// Creating the array of string which are then converted in to an array of interface which are to
 	// be added as header to each of the above spreadsheets.
 	// Example: | First Name | Last Name | Birth Day | Item Purchased | ..
 	// Here messageId is by default the first column
 	headerRowStr = append(headerRowStr, "messageId")
-	for _, eventmap := range config.EventKeyMap {
+	for _, eventmap := range p.config.EventKeyMap {
 		headerRowStr = append(headerRowStr, eventmap["to"])
 	}
 	headerRow := getSheetsData(headerRowStr)
 
-	client := &Client{service, o}
-	// *** Adding the header ***
-	// Inserting header to the sheet
-	err = insertHeaderDataToSheet(client, config.SheetId, config.SheetName, headerRow)
-
-	return &GoogleSheetsProducer{client}, err
+	err := p.insertHeaderDataToSheet(headerRow)
+	if err != nil {
+		return err
+	}
+	p.isHeaderUpdated = true
+	return nil
 }
 
-func (producer *GoogleSheetsProducer) Produce(jsonData json.RawMessage, _ interface{}) (statusCode int, respStatus, responseMessage string) {
-	client := producer.client
-	if client == nil {
+func (p *GoogleSheetsProducer) Produce(jsonData json.RawMessage, _ interface{}) (statusCode int, respStatus, responseMessage string) {
+	if p.client == nil {
 		respStatus = "Failure"
 		responseMessage = "[GoogleSheets] error  :: Failed to initialize google-sheets client"
 		return 400, respStatus, responseMessage
 	}
+	if err := p.updateHeader(); err != nil {
+		statCode, serviceMessage := handleServiceError(err)
+		respStatus = "Failure"
+		responseMessage = "[GoogleSheets] error :: Failed to update header :: " + serviceMessage
+		pkgLogger.Errorn("[Googlesheets] error while updating header", obskit.Error(err))
+		return statCode, respStatus, responseMessage
+	}
+
```

Comment:
Since this is in produce method how do you know wether the config has changed and you need to update the header?

Reply:
we don't need that as  we whole producer will be recreated